File: docs/src/test/java/org/springframework/security/docs/features/authentication/authenticationcompromisedpasswordcheck/CompromisedPasswordCheckerUsage.java
package org.springframework.security.docs.features.authentication.authenticationcompromisedpasswordcheck;

import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import org.springframework.context.annotation.Bean;
import org.springframework.security.authentication.password.CompromisedPasswordChecker;
import org.springframework.security.authentication.password.CompromisedPasswordException;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.DefaultRedirectStrategy;
import org.springframework.security.web.RedirectStrategy;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.AuthenticationFailureHandler;
import org.springframework.security.web.authentication.SimpleUrlAuthenticationFailureHandler;
import org.springframework.security.web.authentication.password.HaveIBeenPwnedRestApiPasswordChecker;

public class CompromisedPasswordCheckerUsage {
	// tag::configuration[]
	@Bean
	public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
		http
			.authorizeHttpRequests(authorize -> authorize
				.anyRequest().authenticated()
			)
			.formLogin((login) -> login
				.failureHandler(new CompromisedPasswordAuthenticationFailureHandler())
			);
		return http.build();
	}

	@Bean
	public CompromisedPasswordChecker compromisedPasswordChecker() {
		return new HaveIBeenPwnedRestApiPasswordChecker();
	}

	static class CompromisedPasswordAuthenticationFailureHandler implements AuthenticationFailureHandler {

		private final SimpleUrlAuthenticationFailureHandler defaultFailureHandler = new SimpleUrlAuthenticationFailureHandler(
				"/login?error");

		private final RedirectStrategy redirectStrategy = new DefaultRedirectStrategy();

		@Override
		public void onAuthenticationFailure(HttpServletRequest request, HttpServletResponse response,
				AuthenticationException exception) throws IOException, ServletException {
			if (exception instanceof CompromisedPasswordException) {
				this.redirectStrategy.sendRedirect(request, response, "/reset-password");
				return;
			}
			this.defaultFailureHandler.onAuthenticationFailure(request, response, exception);
		}

	}
	// end::configuration[]
}

File: docs/src/test/kotlin/org/springframework/security/kt/docs/features/authentication/authenticationcompromisedpasswordcheck/CompromisedPasswordCheckerUsage.kt
package org.springframework.security.kt.docs.features.authentication.authenticationcompromisedpasswordcheck

import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.context.annotation.Bean
import org.springframework.security.authentication.password.CompromisedPasswordChecker
import org.springframework.security.authentication.password.CompromisedPasswordException
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.invoke
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.DefaultRedirectStrategy
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.AuthenticationFailureHandler
import org.springframework.security.web.authentication.SimpleUrlAuthenticationFailureHandler
import org.springframework.security.web.authentication.password.HaveIBeenPwnedRestApiPasswordChecker


open class CompromisedPasswordCheckerUsage {
    // tag::configuration[]
    @Bean
    open fun filterChain(http: HttpSecurity): SecurityFilterChain {
        http {
            authorizeHttpRequests {
                authorize(anyRequest, authenticated)
            }
            formLogin {
                authenticationFailureHandler = CompromisedPasswordAuthenticationFailureHandler()
            }
        }
        return http.build()
    }

    @Bean
    open fun compromisedPasswordChecker(): CompromisedPasswordChecker {
        return HaveIBeenPwnedRestApiPasswordChecker()
    }

    class CompromisedPasswordAuthenticationFailureHandler : AuthenticationFailureHandler {
        private val defaultFailureHandler = SimpleUrlAuthenticationFailureHandler("/login?error")
        private val redirectStrategy = DefaultRedirectStrategy()

        override fun onAuthenticationFailure(
            request: HttpServletRequest,
            response: HttpServletResponse,
            exception: AuthenticationException
        ) {
            if (exception is CompromisedPasswordException) {
                redirectStrategy.sendRedirect(request, response, "/reset-password")
                return
            }
            defaultFailureHandler.onAuthenticationFailure(request, response, exception)
        }
    }
    // end::configuration[]
}
